ledger/common/address.go

```diff
@@ -107,24 +107,47 @@ func NewAddressFromParts(
 	paymentAddr []byte,
 	stakingAddr []byte,
 ) (Address, error) {
+	// Validate network ID
+	if networkId != AddressNetworkTestnet && networkId != AddressNetworkMainnet {
+		return Address{}, errors.New("invalid network ID")
+	}
+
+	// Handle stake-only addresses
+	if addrType == AddressTypeNoneKey || addrType == AddressTypeNoneScript {
+		if len(paymentAddr) > 0 {
+			return Address{}, errors.New("payment address must be empty for stake-only addresses")
+		}
+		if len(stakingAddr) != AddressHashSize {
+			return Address{}, fmt.Errorf("staking key must be exactly %d bytes", AddressHashSize)
+		}
+
+		if addrType == AddressTypeNoneScript && networkId == AddressNetworkTestnet {
+			header := byte(0xF1)
+			addrBytes := append([]byte{header}, stakingAddr...)
+			return NewAddressFromBytes(addrBytes)
+		}
+
+		header := addrType<<4 | networkId
+		addrBytes := append([]byte{header}, stakingAddr...)
+		return NewAddressFromBytes(addrBytes)
```

Comment:
Use the method that we already had for building a new `Address` rather than round-tripping it through bytes